src/settings/settings_manager.cpp

```diff
@@ -0,0 +1,175 @@
+
+using Index = catalog::SettingsTableColumn;
+using ValueFactory = type::ValueFactory;
+
+// Used for building temporary transactions
+void EmptyCallback(void * /*unused*/) {}
+
+SettingsManager::SettingsManager(const std::shared_ptr<catalog::Catalog> &catalog,
+                                 transaction::TransactionManager *txn_manager)
+    : settings_handle_(catalog->GetSettingsHandle()), txn_manager_(txn_manager) {
+  InitParams();
+  InitializeCatalog();
+}
+
+void SettingsManager::InitParams() {
+// This will expand to invoke settings_manager::DefineSetting on
+// all of the settings defined in settings.h. See settings_common.h.
+#define __SETTING_DEFINE__
+#include "settings/settings_common.h"  // NOLINT
+#include "settings/settings_defs.h"    // NOLINT
+#undef __SETTING_DEFINE__
+}
+
+void SettingsManager::DefineSetting(Param param, const std::string &name, const type::Value &value,
+                                    const std::string &description, const type::Value &default_value,
+                                    const type::Value &min_value, const type::Value &max_value, bool is_mutable,
+                                    callback_fn callback) {
+  if (value.Type() == type::TypeId::INTEGER || value.Type() == type::TypeId::DECIMAL) {
+    if (!value.CompareBetweenInclusive(min_value, max_value)) {
+      SETTINGS_LOG_ERROR(
+          "Value given for \"{}"
+          "\" is not in its min-max bounds ({}-{})",
+          name, min_value.PeekAsString(), max_value.PeekAsString());
+      throw SETTINGS_EXCEPTION("Invalid setting value");
+    }
+  }
+
+  param_map_.emplace(param, ParamInfo(name, value, description, default_value, is_mutable));
+  callback_map_.emplace(param, callback);
+}
+
+void SettingsManager::InitializeCatalog() {
+  auto txn = txn_manager_->BeginTransaction();
+  auto column_num = catalog::SettingsHandle::schema_cols_.size();
+
+  for (const auto &pair : param_map_) {
+    Param param = pair.first;
+    ParamInfo info = pair.second;
+
+    catalog::settings_oid_t oid(static_cast<uint32_t>(param));
+    std::vector<type::Value> entry(column_num, type::ValueFactory::GetNullValue(type::TypeId::VARCHAR));
+
+    entry[static_cast<int>(Index::OID)] = ValueFactory::GetIntegerValue(!oid);
+    entry[static_cast<int>(Index::NAME)] = ValueFactory::GetVarcharValue(info.name.c_str());
+    entry[static_cast<int>(Index::SHORT_DESC)] = ValueFactory::GetVarcharValue(info.desc.c_str());
+
+    settings_handle_.InsertRow(txn, entry);
+  }
+
+  txn_manager_->Commit(txn, EmptyCallback, nullptr);
+  if(!txn_manager_->GCEnabled())
+    delete txn;
+}
+
+int32_t SettingsManager::GetInt(Param param) { return GetValue(param).GetIntValue(); }
+
+double SettingsManager::GetDouble(Param param) { return GetValue(param).GetDecimalValue(); }
+
+bool SettingsManager::GetBool(Param param) { return GetValue(param).GetBooleanValue(); }
+
+std::string SettingsManager::GetString(Param param) { return GetValue(param).GetVarcharValue(); }
```

Comment:
Be careful here. If you are using the new value system, there might be lifetime issues. Consider where to make a copy and where to just "peek" values.

Reply:
Thanks for your reminder. We don't make copies here in the new TransientValue System.